dss/storage/validator.py

```diff
@@ -1,23 +1,91 @@
 """Utilities in this file are used to removed extra fields from index data before adding to the index."""
+import io
 import re
+
+import boto3
+from cloud_blobstore import BlobNotFoundError
 from jsonschema import validators
 from jsonschema import _utils, _validators
 
 import json
-from urllib.request import urlopen
-from urllib.error import URLError
-import ssl
-from typing import List
+from typing import List, Optional
 import logging
 
+from requests import RequestException
+
+from dss import Replica, Config, DSSException
+from hashlib import sha1
+import requests
+
+
+DSS_Draft4Validator = validators.create(meta_schema=_utils.load_schema("draft4"),
+                                        validators={u"$ref": _validators.ref,
+                                                    u"additionalProperties": _validators.additionalProperties,
+                                                    u"properties": _validators.properties_draft4,
+                                                    u"required": _validators.required_draft4
+                                                    },
+                                        version="draft4"
+                                        )
+
+
+class SizeLimitError(requests.RequestException):
+    "The max size content size has been exceeded"
+    pass
+
+class S3UrlCache:
```

Comment:
I think you should replace this whole class with a function (maybe `caching_get_url(...)`?) with a `functools.lru_cache()` decorator.

Reply:
I'm confused by this request. This is a persistent cache, not just an in-memory one.

Reply:
Why not make it both, and simplify the interface?
```
@functools.lru_cache()
def caching_get_url(url):
    try:
        requests.get(... from cache ...)
    except ...:
        requests.get(... from original url ...)
        # save to S3
    return content
```

Reply:
It would mix two separate concerns. A user of the S3 Cache might not want the memory overhead that `lru_cache` entails. If they do want in-memory caching they can easily write the above function themselves.

Furthermore, In this particular use case of the cache–resolving JSON schema URLs–an in memory cache is already provided by the RefResolver as @Bento007 pointed out to me yesterday.

Reply:
I still think it would be better to use a single function with an in-memory LRU cache provided by the standard library.

Reply:
Marking as approved to unblock the PR since it looks OK otherwise.